File: listings/cl-existence-oriented/optional-arguments.zig
const std = @import("std");
const allocator = std.heap.page_allocator;

pub fn main() void {
    printOptionIfExists(null);
    printOptionIfExists(42);
    //formatTitle("Hello, world!", null, null);
}

fn printOptionIfExists(option: ?i32) void {
    if (option) |data| {
        std.debug.print("{}\n", .{data});
    }
}

// TODO: Finish or use https://github.com/suirad/zig-strings/blob/master/strings.zig.
// Reference: https://ziglang.org/documentation/master/std/#std;ArrayList
//fn formatTitle(
//    base: []const u8, variant: ?[]const u8, version: ?[]const u8
//) std.mem.Allocator.Error![]u8 {
//    var output = std.ArrayList([]const u8).init(allocator);
//    try output.append(base);
//    if (variant) |data| {
//        try output.append(data);
//    }
//    if (version) |data| {
//        try output.append(data);
//    }
//}

File: listings/cl-existence-oriented/optional-arguments.kt
import platform.posix.*

val stderr = fdopen(2, "w")

fun main() {
    eprintln(format_title("Hello, world!"))
    eprintln(format_title("Hello, world!", variant = "Kotlin"))
    eprintln(format_title("Hello, world!", version = "1.0"))
    eprintln(format_title(
        "Hello, world!", variant = "Kotlin", version = "1.0"
    ))
}

fun eprintln(message: String) {
    fprintf(stderr, message + "\n")
    fflush(stderr)
}

fun format_title(
    base: String, variant: String? = null, version: String? = null
): String {
    val output = mutableListOf(base)
    if (variant != null) {
        variant.length  // Compilation error if not wrapped in null check.
        output.add("[$variant]")
    }
    if (version != null) {
        output.add("($version)")
    }
    return output.joinToString(separator = " ")
}
